SYCL/ESIMD/bfn.cpp

```diff
@@ -0,0 +1,272 @@
+    }                                                                          \
+  };
+
+DEFINE_ESIMD_DEVICE_OP(F1);
+DEFINE_ESIMD_DEVICE_OP(F2);
+DEFINE_ESIMD_DEVICE_OP(F3);
+
+// --- Generic kernel calculating a binary function operation on array elements.
+
+template <class T, int N, experimental::esimd::bfn_t Op,
+          template <class, int, experimental::esimd::bfn_t, int> class Kernel,
+          typename AccIn, typename AccOut>
+struct DeviceFunc {
+  AccIn In0;
+  AccIn In1;
+  AccIn In2;
+  AccOut Out;
+
+  DeviceFunc(AccIn &In0, AccIn &In1, AccIn &In2, AccOut &Out)
+      : In0(In0), In1(In1), In2(In2), Out(Out) {}
+
+  void operator()(id<1> I) const SYCL_ESIMD_KERNEL {
+    unsigned int Offset = I * N * sizeof(T);
+    esimd::simd<T, N> V0;
+    esimd::simd<T, N> V1;
+    esimd::simd<T, N> V2;
+    V0.copy_from(In0, Offset);
+    V1.copy_from(In1, Offset);
+    V2.copy_from(In2, Offset);
+
+    if (I.get(0) % 2 == 0) {
+      for (int J = 0; J < N; J++) {
+        Kernel<T, N, Op, AllSca> DevF{};
+        T Val0 = V0[J];
+        T Val1 = V1[J];
+        T Val2 = V2[J];
+        esimd::simd<T, N> V = DevF(Val0, Val1, Val2); // scalar arg
+        V0[J] = V[J];
+      }
+    } else {
+      Kernel<T, N, Op, AllVec> DevF{};
+      V0 = DevF(V0, V1, V2); // vector arg
+    }
+    V0.copy_to(Out, Offset);
+  };
+};
+
+// --- Generic test function for boolean function.
+
+template <class T, int N, experimental::esimd::bfn_t Op, int Range,
+          template <class, int, experimental::esimd::bfn_t, int> class Kernel,
+          typename InitF = InitOps<T>>
+bool test(queue &Q, const std::string &Name, InitF Init = InitOps<T>{}) {
+  constexpr size_t Size = Range * N;
+
+  T *A = new T[Size];
+  T *B = new T[Size];
+  T *C = new T[Size];
+  T *D = new T[Size];
+  Init(A, B, C, D, Size);
+
+  std::cout << "  " << Name << " test"
+            << "...\n";
+
+  try {
+    buffer<T, 1> BufA(A, range<1>(Size));
+    buffer<T, 1> BufB(B, range<1>(Size));
+    buffer<T, 1> BufC(C, range<1>(Size));
+    buffer<T, 1> BufD(D, range<1>(Size));
+
+    // number of workgroups
+    sycl::range<1> GlobalRange{Range};
+
+    // threads (workitems) in each workgroup
+    sycl::range<1> LocalRange{1};
+
+    auto E = Q.submit([&](handler &CGH) {
+      auto PA = BufA.template get_access<access::mode::read>(CGH);
+      auto PB = BufB.template get_access<access::mode::read>(CGH);
+      auto PC = BufC.template get_access<access::mode::read>(CGH);
+      auto PD = BufD.template get_access<access::mode::write>(CGH);
+      DeviceFunc<T, N, Op, Kernel, decltype(PA), decltype(PD)> F(PA, PB, PC,
+                                                                 PD);
+      CGH.parallel_for(nd_range<1>{GlobalRange, LocalRange}, F);
+    });
+    E.wait();
+  } catch (sycl::exception &Exc) {
+    std::cout << "    *** ERROR. SYCL exception caught: << " << Exc.what()
+              << "\n";
+    return false;
```

Comment:
This line does return without freeing memory allocated above. Such test may report fails in some special modes monitoring memory leaks.

Reply:
One of easiest ways to allocate memory and let it be released automatically on exit from function is this one: https://github.com/intel/llvm-test-suite/blob/intel/SYCL/ESIMD/imulh_umulh.cpp#L64-L69

Reply:
Done.